hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java

```diff
@@ -663,36 +678,68 @@ private ContainerMoveSelection matchSourceWithTarget(DatanodeDetails source) {
     return moveSelection;
   }
 
+  private boolean reachedMaxSizeToMovePerIteration() {
+    // since candidate containers in ContainerBalancerSelectionCriteria are
+    // filtered out according to this limit, balancer should not have crossed it
+    if (sizeScheduledForMoveInLatestIteration >= maxSizeToMovePerIteration) {
+      LOG.warn("Reached max size to move limit. {} bytes have already been" +
+              " scheduled for balancing and the limit is {} bytes.",
+          sizeScheduledForMoveInLatestIteration, maxSizeToMovePerIteration);
+      return true;
+    }
+    return false;
+  }
+
   /**
-   * Checks if limits maxDatanodesPercentageToInvolvePerIteration and
-   * maxSizeToMovePerIteration have been hit.
-   *
-   * @return true if a limit was hit, else false
+   * Restricts potential target datanodes to nodes that have
+   * already been selected if balancer is one datanode away from
+   * "datanodes.involved.max.percentage.per.iteration" limit.
+   * @return true if potential targets were restricted, else false
    */
-  private boolean checkIterationLimits() {
+  private boolean adaptWhenNearingIterationLimits() {
+    // check if we're nearing max datanodes to involve
     int maxDatanodesToInvolve =
         (int) (maxDatanodesRatioToInvolvePerIteration * totalNodesInCluster);
-    if (countDatanodesInvolvedPerIteration + 2 > maxDatanodesToInvolve) {
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("Hit max datanodes to involve limit. {} datanodes have" +
-                " already been scheduled for balancing and the limit is {}.",
-            countDatanodesInvolvedPerIteration, maxDatanodesToInvolve);
-      }
+    if (countDatanodesInvolvedPerIteration + 1 == maxDatanodesToInvolve) {
+      /* We're one datanode away from reaching the limit. Restrict potential
+      targets to targets that have already been selected.
+       */
+      findTargetStrategy.resetPotentialTargets(selectedTargets);
```

Comment:
if we reach here, i think we can restrict any of potential targets or sources, as long as we make sure we will involve at most one more datanode.

for example , when all the selected targets can not be a candidate target again(they may be overutilized after receiving a container replica) , we can restrict potential sources to sources that have already been selected, so that we may select a new target which will not be overutilized after receiving a container replica, meanwhile we involve only one more datanode in this balance iteration.

Reply:
Yes, that's a good point. Let me see how to do that.

Reply:
Looking into it, I don't see a straightforward way to do this. @JacksonYao287 Since this PR has been open for quite a while, do you think we should create a different JIRA to investigate restricting sources when targets are unavailable?

Reply:
agree, we can create another jira to do this.
@lokeshj1703 do you  have any comments?